### PR TITLE
Adds TokenStreamReader

### LINKED ISSUE #520: API for Efficient Stream of Ion Values
We have an `IonReader` trait that captures both the *"iteration"* over the ***stream*** and the ***value*** the stream is currently on.  This is a fine API, but there are cases (especially when we start to think about macros in Ion 1.1), that this API becomes difficult to compose on.  I propose we create an API that de-couples the ***stream*** from the ***value*** without while deferring computation on the value it is on.

At a high-level, the API might look something like this:

```rs
pub enum Instruction {
  /// Advance to the next token.
  Next,
  /// Skip to the end of the current container
  NextContainerEnd,
}

struct AnnotatedToken<'a>{ ... }

trait TokenStream {
  fn next_token(&mut self, instruction: Instruction) -> IonResult<AnnotatedToken>;
}
```

The above is quite a simplification, but it demonstrates the basic structure.  `AnnotatedToken` would represent the next token deferring computation on a field name, annotations, or the underlying scalar value, the beginning of a container, end of the container, or the end of a stream.

It would be straightforward to adapt this `TokenStream` wrap any `IonReader` (though some care would have to be done to model this on both user-level/system-level/raw-level variants to do it generically), and the other direction should be relatively straightforward as well (make any `TokenStream` into an `IonReader`.   The latter gives us `TokenStream` to `Element`.

### PR DESCRIPTION
**Pull request by almann**
`TokenStreamReader` adapts `TokenStream` into a user-level `Reader`.

This also changes the support types for `ElementReader` to work with `IonReader` trait objects (adding `?Sized` to the type constraints).

The `Element` round-trip tests are also adapted to work with _any_ `IonReader` as a boxed trait object, which allows us to slot in `TokenStreamReader` over `ReaderTokenStream` over a user `Reader` instance, exercising both adapters in this framework.

This also consolidates the round-trip skip-lists into constants so there is less copy/paste between the now three instances of the `ElementApi` trait.

resolves #520

----

Note that the this is dependent on #541 and currently based off of that--when/if that is merged **and** this is approved I will change to base to main.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/tokens/stream_reader.rs [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+/// Generates the read methods against the underlying token/item state.
+/// This eliminates the boilerplate around each definition avoiding a lot of copy/paste.
+///
+/// * `me` - the identifier for `self` to make it visible to `scalar_exp`.
+/// * `method` - the name of the method.
+/// * `scalar_type` - the return type for the method.
+/// * `variant` - the unqualified variant of [`ScalarValue`] to match over the current token.
+/// * `scalar` - the matched identifier of the contents of [`ScalarValue`].
+/// * `scalar_exp` - the expression to generate the return value.
+macro_rules! read_method_self {
```
Comment (almann) [inline src/tokens/stream_reader.rs:93]: 🗺️ It makes me sad to do this, but this macro really does avoid a lot of boilerplate around implementing the `read_XXX` methods especially given our `Option<RefCell<Token>>` modalities.

File: src/tokens/stream_reader.rs [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+    T: TokenStream<'a>,
+{
+    type Item = StreamItem;
+    type Symbol = Symbol;
+
+    fn ion_version(&self) -> (u8, u8) {
+        // XXX A `TokenStream` doesn't know its underlying Ion version, we just say 1.0
+        //     because all Ion 1.x versions have the same data model.
+        (1, 0)
+    }
```
Comment (almann) [inline src/tokens/stream_reader.rs:133]: 🗺️ `ElementStreamReader` has a similar weirdness--I honestly don't think this API should be here, it probably belongs at a lower-level or some other trait.

File: src/tokens/stream_reader.rs [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+        };
+        Box::new(iter.map(Ok))
+    }
+
+    fn field_name(&self) -> IonResult<Self::Symbol> {
+        match &self.curr_token_cell {
+            None => illegal_operation(NO_FIELD_NAME_ERROR_TEXT),
+            Some(token_cell) => {
+                let mut token = token_cell.borrow_mut();
+                match token.share_field_name() {
```
Comment (almann) [inline src/tokens/stream_reader.rs:214]: 🗺️ We could be more aggressive about _not_ interning, only intern when this is called after we materialized it once, but I don't think that is worth it, especially while we're re-thinking the `IonReader` interface in #536.

File: tests/element_test_vectors.rs [hunk 4]
```diff
@@ -419,83 +422,90 @@ mod impl_display_for_element_tests {
     }
 }
 
+const ELEMENT_GLOBAL_SKIP_LIST: SkipList = &[
```
Comment (almann) [inline tests/element_test_vectors.rs:425]: 🗺️ It appears that these are identical across all `ElementApi` implementations--I'd bet this was a vestige from when we had Ion C lurking.

File: tests/element_test_vectors.rs [hunk 5]
```diff
@@ -67,14 +67,17 @@ fn serialize(format: Format, elements: &[Element]) -> IonResult<Vec<u8>> {
     Ok(buffer)
 }
 
+/// Allow for anything that implements a user-like reader
+type DynamicReader<'a> = Box<dyn IonReader<Item = StreamItem, Symbol = Symbol> + 'a>;
```
Comment (almann) [inline tests/element_test_vectors.rs:71]: 🗺️ This lets us plug in the `ElementStreamReader` if we want to which is probably not a bad idea for the future (more coverage).

File: tests/element_test_vectors.rs [hunk 6]
```diff
@@ -665,3 +628,77 @@ mod non_blocking_native_element_tests {
...
+        }
+
+        fn non_equivs_skip_list() -> SkipList {
+            &[]
+        }
+
+        fn make_reader<'a>(data: &'a [u8]) -> IonResult<DynamicReader<'a>> {
+            let reader = ReaderBuilder::default().build(data)?;
+            let tokens: ReaderTokenStream<'a, _> = reader.into();
+            let token_reader: TokenStreamReader<'a, _> = tokens.into();
```
Comment (almann) [inline tests/element_test_vectors.rs:664]: 🗺️ I think this gives us a lot of mileage for testing both adapters.

**Review by almann — COMMENTED** [review]
🗺️ PR tour below.

File: src/tokens/stream_reader.rs [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+        //     because all Ion 1.x versions have the same data model.
+        (1, 0)
+    }
+
+    fn next(&mut self) -> IonResult<Self::Item> {
+        match &self.curr_token_cell {
+            None => self.next_token(Instruction::Next),
+            Some(token_cell) => {
+                let token = token_cell.borrow();
+                let insn_opt: Option<_> = match token.content() {
```
Comment (zslayton) [inline src/tokens/stream_reader.rs:140]: ~What does `insn` mean?~

EDIT: I see, it's "instruction". I think this variable name only appears twice in the method, it'd be worth spelling it out.

**Review by zslayton — APPROVED** [review]

**Pull request merged**